x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/common/ProgressListenableActionFuture.java

```diff
@@ -26,12 +28,17 @@ class ProgressListenableActionFuture extends PlainActionFuture<Long> {
 
     private record PositionAndListener(long position, ActionListener<Long> listener) {}
 
-    protected final long start;
-    protected final long end;
+    final long start;
+    final long end;
 
-    // modified under 'this' mutex
-    private volatile List<PositionAndListener> listeners;
-    protected volatile long progress;
+    /**
+     * A consumer that accepts progress made by this {@link ProgressListenableActionFuture}
```

Comment:
Can we add this to the constructor javadoc too. And also mention that this is called prior to notifying listeners (since we'll rely on that), though with out of order possibilities for concurrent notifications.

Reply:
I pushed b15d35150e3dd0d17a59daf908524c1c4ccffd76